game/groundforces/template_loader.py

```diff
@@ -0,0 +1,361 @@
+        with file.open("wb") as fdata:
+            pickle.dump(dump, fdata)
+
+    @staticmethod
+    def mapping_for_group(
+        mappings: list[TemplateMapping], group_name: str
+    ) -> tuple[TemplateMapping, int, GroupTemplateMapping]:
+        for mapping in mappings:
+            for g_id, group_mapping in enumerate(mapping.groups):
+                if (
+                    group_mapping.name == group_name
+                    or group_name in group_mapping.statics
+                ):
+                    return mapping, g_id, group_mapping
+        raise KeyError
+
+    def _load_from_miz(self, miz: str, mappings: list[TemplateMapping]) -> None:
+        template_position: dict[str, Point] = {}
+        temp_mis = dcs.Mission()
+        with logged_duration("dcs mission loading"):
+            # The load_file takes a lot of time to compute. That's why the templates
+            # are written to a pickle and can be reloaded from the ui
+            # Example the whole routine: 0:00:00.934417,
+            # the .load_file() method: 0:00:00.920409
+            temp_mis.load_file(miz)
+
+        for country in itertools.chain(
+            temp_mis.coalition["red"].countries.values(),
+            temp_mis.coalition["blue"].countries.values(),
+        ):
+            for dcs_group in itertools.chain(
+                temp_mis.country(country.name).vehicle_group,
+                temp_mis.country(country.name).ship_group,
+                temp_mis.country(country.name).static_group,
+            ):
+                try:
+                    mapping, group_id, group_mapping = self.mapping_for_group(
+                        mappings, dcs_group.name
+                    )
+                except KeyError:
+                    logging.error(f"No mapping for dcs group {dcs_group.name}")
+                    continue
+
+                template = self._templates.get(mapping.name, None)
+                if template is None:
+                    # Create a new template
+                    template = TEMPLATE_TYPES[mapping.role](
+                        mapping.name, mapping.role, mapping.description
+                    )
+                    template.generic = mapping.generic
+                    template.tasks = mapping.tasks
+                    self._templates[template.name] = template
+
+                for i, unit in enumerate(dcs_group.units):
+                    group_template = None
+                    for group in template.groups:
+                        if group.name == dcs_group.name or (
+                            isinstance(dcs_group, StaticGroup)
+                            and dcs_group.units[0].type in group.unit_types
+                        ):
+                            # MovingGroups are matched by name, statics by unit_type
+                            group_template = group
+                    if not group_template:
+                        group_template = GroupTemplate(
```

Comment:
Oh. This explains a lot of my confusion elsewhere. I did not expect a copy of a template to be made for every group that uses it. That seems way over complicated, and also explains why template loading is so damn expensive.

Why? I haven't seen anything that would make the copies necessary. They're pretty simple static data. The interesting per-group mutations can happen when the template is _used_.